Static templates: also support ``.jinja`` files

jayaddison opened the pull request:
### Feature or Bugfix
- Feature

### Purpose
- Allow themes and other extensions that utilise the static template functionality to use the [suggested](https://jinja.palletsprojects.com/en/3.1.x/templates/#template-file-extension) (note: not strictly required) `.jinja` extension for templates.
- The originating motivation I have for this feature is to eventually remove a catch-all missing-suffix expectation in `coveragepy` that can cause warnings to be emitted when non-Python files are treated as potentially containing Python bytecode for coverage evaluation (ref https://github.com/nedbat/coveragepy/issues/1488).

### Detail
- When considering whether to render a file using Jinja, check for _either_ the existing already-supported `_t` suffix _or_ secondly for the additionally-supported `.jinja` suffix.
- These changes currently target version 7.3 of Sphinx, but I'd be fine with waiting until a future version to introduce them.

### Relates
- Re-attempts previous pull request #11165 - this time without deprecation notices or a deprecation schedule.

jayaddison committed:
static templates: also support filenames with .jinja suffix

jayaddison committed:
tests: backwards-compatibility: add test coverage for rendering the contents of a legacy-suffix template

jayaddison committed:
nitpick: do not introduce change-of-behaviour with regard to '_T' suffix template detection on case-sensitive filesystems

jayaddison committed:
ruff: apply suggested linter change to place docstring closing-quotes on separate line

jayaddison committed:
fixup: update a few lingering docstring and message variable references to '_t'-suffix templates

jayaddison committed:
change-isolation: undo template file renames

jayaddison committed:
Revert "fixup: update a few lingering docstring and message variable references to '_t'-suffix templates"

This reverts commit ca21989f7ec16d69d40ca0240c607ff8089fc59e.

jayaddison committed:
change-isolation: undo template file extension coupling in application and test code

Relates-to commits 7518c1439a275546b141cab229329858388b246e, f0f22fd596031bf426fdccd652b564e7ef040633.

jayaddison commented on sphinx/writers/latex.py:
```diff
@@ -517,12 +517,13 @@ def generate(content: list[tuple[str, list[IndexEntry]]], collapsed: bool) -> No
+                    return renderer.render(template, variables)
 
-        return renderer.render(template_name, variables)
+        return renderer.render(template_name + '_t', variables)
```
Something about this line feels wrong to me, in the sense of introducing an unintended behaviour change / incompatibility.

jayaddison replied:
I think it's safest to undo this and further reduce the size of the diff.

jayaddison committed:
change-isolation: undo modifications to LaTeX, imgmath writer template detection and rendering

Relates-to commits cef1059d78d77111dc4a64336b9f9069d99fff11, 7518c1439a275546b141cab229329858388b246e, f0f22fd596031bf426fdccd652b564e7ef040633.

jayaddison committed:
tests: repurpose statictmpl.html template from legacy-suffix to recommended-suffix, providing a forwards-compatibility test

jayaddison committed:
Merge branch 'master' into issue-11165/support-jinja-suffix

jayaddison committed:
Merge branch 'master' into issue-11165/support-jinja-suffix

AA-Turner committed:
Merge branch 'master' into issue-11165/support-jinja-suffix

AA-Turner committed:
Merge branch 'refs/heads/master' into issue-11165/support-jinja-suffix

AA-Turner committed:
Updates

AA-Turner commented on CHANGES.rst:
```diff
@@ -75,6 +75,10 @@ Deprecated
 Features added
 --------------
+* #11165: Support the `officially recommended`_ ``.jinja`` suffix for template
+  files. Patch by James Addison and Adam Turner
```
```suggestion
  files. 
  Patch by James Addison and Adam Turner
```

AA-Turner committed:
Update CHANGES.rst

jayaddison commented on sphinx/util/fileutil.py:
```diff
@@ -23,7 +23,7 @@ def _template_basename(filename: str | os.PathLike[str]) -> str | None:
     basename = os.path.basename(filename)
     if basename.lower().endswith('_t'):
         return str(filename)[:-2]
-    elif basename.lower().endswith(".jinja"):
```
I'm not sure about the removal of `.lower()` here; what's the thinking behind that?

jayaddison replied:
@AA-Turner Although it might seem non-ideal in some ways to allow people to create template files named arbitrary combinations of `[Jj][Ii][Nn][Jj][Aa]`, initially I still feel that it's better to compare the file suffix case-insensitively.

Ref: https://www.hanselman.com/blog/git-is-casesensitive-and-your-filesystem-may-not-be-weird-folder-merging-on-windows for example.

jayaddison committed:
[docs] Theming: update versionchanged record for Jina-suffix support.

jayaddison committed:
Merge branch 'master' into issue-11165/support-jinja-suffix

jayaddison committed:
util.fileutil: restore case-insensitive .jinja suffix filename matches.

Some filesystems, such as APFS, use case-sensitive filenames, but require files to be named distinctly in a case-insensitive manner.  Without using case-insensitive filename comparison, we risk the possibility that, for example, someone using Linux/Windows could create a template file that works fine on their machine, but that fails when someone else tries to build that project on MacOS.

AA-Turner committed:
Merge branch 'master' into issue-11165/support-jinja-suffix

AA-Turner committed:
Merge branch 'refs/heads/master' into issue-11165/support-jinja-suffix

AA-Turner committed:
CHANGES
